drivers/at86rf2xx/at86rf2xx.c

```diff
@@ -32,59 +35,215 @@
 #include "at86rf2xx_registers.h"
 #include "at86rf2xx_internal.h"
 #include "at86rf2xx_netdev.h"
+#include "at86rf2xx_dev_types.h"
+#include "at86rf2xx_properties.h"
+#include "at86rf212b.h"
+#include "at86rf231.h"
+#include "at86rf233.h"
+#include "at86rf233.h"
+#include "at86rfa1.h"
+#include "at86rfr2.h"
 
 #define ENABLE_DEBUG (0)
 #include "debug.h"
 
-
-void at86rf2xx_setup(at86rf2xx_t *dev, const at86rf2xx_params_t *params)
+static void at86rf2xx_setup(at86rf2xx_t *dev)
 {
+    assert(dev->base.dev_type < AT86RF2XX_DEV_TYPE_NUM_OF);
     netdev_t *netdev = (netdev_t *)dev;
-
     netdev->driver = &at86rf2xx_driver;
     /* State to return after receiving or transmitting */
-    dev->idle_state = AT86RF2XX_STATE_TRX_OFF;
+    dev->base.idle_state = AT86RF2XX_STATE_TRX_OFF;
     /* radio state is P_ON when first powered-on */
-    dev->state = AT86RF2XX_STATE_P_ON;
-    dev->pending_tx = 0;
+    dev->base.state = AT86RF2XX_STATE_P_ON;
+    dev->base.pending_tx = 0;
+
+    switch (dev->base.dev_type) {
+#if IS_USED(MODULE_AT86RFA1)
+        case AT86RF2XX_DEV_TYPE_AT86RFA1: {
+            /* set all interrupts off */
+            at86rf2xx_reg_write(dev, AT86RF2XX_REG__IRQ_MASK, 0x00);
+            break;
+        }
+#endif
+#if IS_USED(MODULE_AT86RFR2)
+        case AT86RF2XX_DEV_TYPE_AT86RFR2: {
+            /* set all interrupts off */
+            at86rf2xx_reg_write(dev, AT86RF2XX_REG__IRQ_MASK, 0x00);
+            break;
+        }
+#endif
+    }
+}
 
-#if defined(MODULE_AT86RFA1) || defined(MODULE_AT86RFR2)
-    (void) params;
-    /* set all interrupts off */
-    at86rf2xx_reg_write(dev, AT86RF2XX_REG__IRQ_MASK, 0x00);
+void at86rf212b_setup(at86rf212b_t *devs, const at86rf212b_params_t *params,
+                      uint8_t num)
+{
+#if IS_USED(MODULE_AT86RF212B)
+    for (int i = 0; i < num; i++) {
+        devs->base.dev_type = AT86RF2XX_DEV_TYPE_AT86RF212B;
+        devs->params = *params;
+        at86rf2xx_setup((at86rf2xx_t *)devs);
+        devs++;
+        params++;
+    }
```

Comment:
Just using `devs[i].base.dev_type = AT86RF2XX_DEV_TYPE_AT86RF212B` is more readable and with that you can drop the lines `devs++` and `params++`. (The compiler will generate the same machine code anyway.)

Also: The cast from the more specific "object" to its parent "object" is not needed. You could just use `at86rf2xx_setup(&devs[i].base)` and have the type safety still in place. (Only when converting from the parent `at86rf2xx_t` to the more specific `at86rf212b_t` requires some magic. You can use the `container_of()` macro provided by `kernel_defines.h` for some limited type safety there: `container_of()` will only work if the more specific struct is indeed a "child object" of the `at86rf2xx_t`.

Reply:
Ok,
so I will use `container_of()` for any cast from the parent to the specific type.

Hmm, if I drop the caset `(at86rf2xx_t*)` in `at86rf2xx_setup(&devs[i].base)` my compile complains.
Because `.base` is of type `at86rf2xx_base_t`. The type `at86rf2xx_t` is `at86rf2xx_base_t` + `at86rf2xx_params_t`.